doc/2/guides/develop-on-kuzzle/3-event-system/index.md

```diff
@@ -0,0 +1,155 @@
+
+Each event has a different payload.  
+The pipe handler function **must return a promise resolving to the received payload**.  
+
+::: info
+When an event has more than one payload then only the first argument of the handler function must be returned. (e.g. [Generic Document Events](/core/2/some-link))
+:::
+
+**Example:** _Changing the result of the [server:now](/core/2/api/controllers/server/now) API action_
+
+```js
+app.pipe.register('server:afterNow', async (request: Request) => {
+  request.result.now = (new Date()).toUTCString()
+
+  return request
+})
+```
+
+::: warning
+As pipes are executed synchronously by Kuzzle, they can increase the execution time of a request.  
+A pipe that takes a long time to execute will generate an alert message in the logs.
+This warning can be configured under the [plugins.pipeWarnTime](/core/2/main-concepts/configuration) configuration key.
+:::
+
+### Aborting a task
+
+When the pipe handler function returns a rejected promise, Kuzzle abort the current task.
+
+If the error is one of the [available default errors](/core/2/some-link) then the response returned to the client will contain the error as is, otherwise the error will be wrapped in a [PluginInternalError](/core/2/some-link).
+
+**Example:** _Limit reading access to documents to their creator_
+```js
+app.pipe.register(
+    'generic:document:beforeGet', 
+    async (documents: Document[], request: Request) => {
+      for (const document of documents) {
+        if (request.context.user._id !== document._source._kuzzle_info.creator) {
+          throw new Error('Unauthorized access')
+        }
+      }
+
+      return documents
+    })
+```
+
+::: info
+[Generic Document Events](/core/2/core/2/some-link) have a payload consisting of two arguments: an array of documents and the original [Request object](/core/2/some-link)
+:::
+
+## Hooks
+
+Kuzzle allows to execute additional tasks upon events with **very precise middleware-like system**.  
+
+Hooks are **functions plugged to events**, called **asynchronously** by Kuzzle, and **receiving information** regarding that event.
+
+![hook workflow](./hooks-workflow.png)
+
+::: info
+In general hooks are used to perform background tasks who may slow the request execution process.
+:::
+
+Examples of hooks usage:
+ - [enrich the request with external information](/core/2/some-link)
+ - [notify user registration](/core/2/some-link)
+
+### Registering a hook
+
+We need to use the [Backend.hook.register](/core/2/some-link) method to register new hooks.   This method takes an event name as its first parameter, followed by the hook handler function.
+
+**Example:** _Use the [pub/sub engine](/core/2/main-concepts/5-realtime-engine#pub-sub) to notify user registration_
+
+```js
+app.hook.register('security:afterCreateRestrictedUser', async (request: Request) => {
+  try {
+    await app.sdk.realtime.publish('admin', 'new-user', request.context.user)
```

Comment:
Wow... This example is very fancy but IMHO a bit difficult to understand. Why not just a simple `app.log`?